pingcap/tidb-operator
# Optimize leader eviction before resize TiKV PVCs

### What problem does this PR solve?


### What is changed and how does it work?


* Add a field `status.tikv.evictLeader[].beginTime` to record eviction begin time.
* Skip evicting leader before resizing if only one TiKV.
* Skip waiting for the leader count of a TiKV store to be 0 before resizing when eviction timeout.
* Check resized pods are not existing in `status.tikv.evictLeader` before resizing a new Pod.

### Code changes


### Tests



Resize one TiKV:
1. Deploy cluster with one TiKV.
2. Scale-up storage size from 200Gi to 300Gi.
3. Check the leader eviction is skipped.
4. Scale-up PVC, and check resize finished.

![image](https://user-images.githubusercontent.com/26425336/178743195-e1545866-753d-4b8b-94c5-7e8d16f1d72d.png)



For the `beginTime`  field:
1. Add anno `tidb.pingcap.com/evict-leader:none` to a TiKV Pod.
2. Check `status.tikv.evictLeader[].beginTime` is existing and not changed.


For eviction timeout and wait to end leader eviction for resized Pod:
1. Deploy cluster with 3 TiKV.
2. Add evict leader scheduler for 2nd 3rd TiKV.
3. Scale-up storage size for TiKV.
4. Skip waiting for store of 1th TiKV to be 0 after the timeout.
5. Scale-up PVC so that 1th TiKV is resized
6. Add `tidb.pingcap.com/evict-leader:none`  to 1th TiKV to mock leader eviction scheduler is not removed.
7. Check that resize is stuck because wait to end leader eviction.

![image](https://user-images.githubusercontent.com/26425336/178751828-37b038f4-586a-4746-a33d-9fff7108cf4b.png)

![image](https://user-images.githubusercontent.com/26425336/178753652-3288cccb-c0be-4165-9789-3188ec62d344.png)



### Side effects


### Related changes


### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Optimize leader eviction before resizing PVCs of TiKV 
```
